src/CarbonAware/src/Extensions/EmissionsDataExtensions.cs

```diff
@@ -87,25 +93,52 @@ public static IEnumerable<EmissionsData> RollingAverage(this IEnumerable<Emissio
         }
     }
 
-    private static EmissionsData AverageOverPeriod(this IEnumerable<EmissionsData> data, DateTimeOffset startTime, DateTimeOffset endTime)
+    /// <summary>
+    /// Finds the average rating of a continuous, chronological set of EmissionsData objects for a given period. 
+    /// </summary>
+    /// <remarks>
+    ///     <para>
+    ///         This method takes into account fractional data points and weights the average accordingly.
+    ///         Take the following (HH:MM, rating) hour-long data points for example: (01:00, 30), (02:00, 100), (03:00, 50)
+    ///     </para>
+    ///     <para>
+    ///         To calculate the average over the time period 01:30 - 03:30, the proportion each point contributes
+    ///         to the whole period must be taken into account.  This proportion is calculated as (utilized time / total duration)
+    ///         For the 2 hour period (01:30-03:30) the proportions would be:
+    ///         (30 * (30m/120m)) + (100 * (60m/120m)) + (50 * (30m/120m)); which simplifies to 
+    ///         (30 * (1/4)) + (100 * (1/2)) + (50 * (1/4)); or
+    ///         (30/4) + (100/2) + (50/4)
+    ///     </para>
+    ///     <para>
+    ///         This method works just as well for simple averages typically calculated as the sum of the values divided by the number of values.
+    ///         (30 + 100 + 50)/3;  We can use the equation above to average the three points over their full 3 hour duration.
+    ///         (30 * (60m/180m)) + (100 * (60m/180m)) + (50 * (60m/180m)); which simplifies to
+    ///         (30 * (1/3)) + (100 * (1/3)) + (50 * (1/3)); or
+    ///         (30/3) + (100/3) + (50/3); or further
+    ///         (30 + 100 + 50)/3 
+    ///     </para>
+    /// </remarks>
+    /// <param name="data">The IEnumerable<EmissionsData> being operated on.</param>
+    /// <param name="startTime">The start time of the data to be averaged.</param>
+    /// <param name="endTime">The end time of the data to be averaged.</param>
+    /// <returns>The average rating of the data for the specified time period</returns>
+    public static double AverageOverPeriod(this IEnumerable<EmissionsData> data, DateTimeOffset startTime, DateTimeOffset endTime)
     {
-        EmissionsData newDataPoint = new EmissionsData()
-        {
-            Time = startTime,
-            Duration = (endTime - startTime),
-            Rating = 0.0,
-            Location = data.First().Location
-        };
+        double rating = 0.0;
+        TimeSpan totalDuration = endTime - startTime;
+        DateTimeOffset? lastEndTime = null;
         foreach (var current in data)
         {
+            lastEndTime = (lastEndTime == null || current.Time == lastEndTime) ? current.Time + current.Duration : throw new InvalidOperationException($"AverageOverPeriod requires continuous chronological data. Previous point covered through {lastEndTime}; Current point starts at {current.Time}."); ;
```

Comment:
My point is that do we want to fail for chronological order, knowing that we can make it work by sorting the data first? This implementation assumes that the provider is a nice citizen giving us the data this way.

Reply:
@juzuluag Ah I see what you are getting at.  The big reason for this approach is simply that most time-series data is sorted, and this method is only relevant to time-series data.  It seemed too much to pay the cost to sort it every time (when it is likely already sorted).  And if the data isn't already chronological and continuous there is no reason to believe that it will become that way when sorted.

In the end, I felt that it made more sense to just assert that the data is sorted and let callers manage putting it in the right state if they have concerns about the ordering.  EG `myList.Sort().AverageOverPeriod(start, end);`

I have updated the docs to highlight this error-case explicitly